### PR TITLE
Interpreting a casted term in the scope bound to its type if any

### LINKED ISSUE #14959: ssreflect pollutes notation precedence space for colon, again
#### Description of the problem

This seems to be #6078 again. VST defines a conflicting notation.

```coq
From Coq Require Import ssreflect.
From VST Require Import floyd.proofauto.
(*
Notation "_ : _" is already defined at level 100 with arguments constr
at next level, constr at level 100 while it is now required to be at level 26
with arguments constr at next level, constr at level 26.
*)
```

#### Coq Version

Coq 8.13

### PR DESCRIPTION
**Pull request by herbelin**
This is a proposal in relation to the discussion at #6078.

It proceeds in several steps depending on taste:
- setting `type_scope` when casted by a sort
- removal of the then-redundant vernac-level notation casts in `ssreflect.v`
- interpreting any term casted with a type bound to a scope in this scope
- a start at interpreting any pattern casted with a type bound to a scope in this scope, as also discussed at #6078.

There are more to do about the last step but that will become easier in the context of the more regular parsing of patterns in #982.

Note 1: I have a terminology question: how to call the `bind` in `Option`? In which order should be its arguments? (I tentatively called it `compose`.)
Note 2 : This PR depends on #6133 (replaced by #18031) and ~#982 (merged)~ shall depend on it for interpreting casts in 
patterns.

Fixes: #14959 

#### Overlays (to be merged before the current PR)

- [x] https://github.com/coq-community/corn/pull/199
- [x] https://github.com/coq-community/math-classes/pull/118

### CODE REVIEW

**Comment by Zimmi48** [comment]
> I have a terminology question: how to call the bind in Option? In which order should be its arguments? (I tentatively called it compose.)

I haven't checked how the other binds are called in the code base currently but I would be in favor of uniformity of naming. For instance, in Elm all the binds are called `andThen`: this particular one is `Maybe.andThen` (`Maybe` being the terminology inherited from Haskell to designate `Option`), and there is also `Result.andThen`, `Task.andThen`, etc.
So if the name `bind` is already used, then it could be used here as well.

As for the order of the arguments, the “standard” order coming from Haskell is to put the argument encapsulated in a monad first, and the function then, but Elm has recently decided to change this order for a very good reason: compatibility with the `|>` (pipe operator). This gives:

```elm
Just 1
|> Maybe.andThen (\x -> if x = 0 then Nothing else x + x) 
|> Maybe.map (\x -> x - 1)
```

But while I like this order of argument better, I think the most important thing is consistency. So if Coq has already got binds with the other order, keep the other order for this new bind as well (unless we change the order for all the binds at once).

**Comment by herbelin** [comment]
> > I have a terminology question: how to call the bind in Option? In which order should be its arguments? (I tentatively called it compose.)
>
> I haven't checked how the other binds are called in the code base currently but I would be in favor of uniformity of naming.

There is already not a lot of uniformity. For instance the `return` in `Option` is `make`.

Anyway, there are many ways to look at an object. Shall we favor a name which reflect some intuition of the object, or another intuition? Shall we reflect algebraic properties of an object, but why such property and not another one.

But I think it is also important to share a common language, and that's why I preventively raised the question.

**Comment by maximedenes** [comment]
A rebase seems needed.

**Comment by herbelin** [comment]
This PR introduces incompatibilities in HoTT and math-classes:
- HoTT: a scope implicitly propagated to the body of a cast is not propagated any more. I guess @JasonGross, who asked for the current PR, is ok for a fix. However, it is unclear what to do. We had `Notation "S / T" := (comma_category (S : CC_Functor' _ _) (T : CC_Functor' _ _))` and thus `S` and `T` in scope `%functor` inherited from `comma_category`. If I add a `%functor` around `S` and `T`, it is too strong as it recursively changes the scope towards the structure of `S` and `T`. One would need a way to say temporarily set `%functor` and there is no syntax for that. Maybe should we provide one?
- math-classes ([diff](https://github.com/herbelin/math-classes/commit/f8fbd645aa741cf90f6fff433301414ff23efb8c)): several instances did not need a cast, so I removed the cast. Other instances were expecting a cast to help type class inferences. I added an explicit `%mc` to enforce the scope as it was formerly. Don't know if it is ok for @spitters.

**Comment by spitters** [comment]
It's a small change. I can live with that for math-classes.
Could you send a PR when this gets merged?

**Comment by Zimmi48** [comment]
Just making sure: is this fix backward compatible?

**Comment by JasonGross** [comment]
What breaks if `%functor` is added in HoTT?

**Comment by JasonGross** [comment]
Looking at the commits to HoTT: oh this is interesting.  I think the thing I actually want here is to be able to bind `category_scope` to `object Cat` (while leaving `object_scope` bound to `object _`), and have bound scopes added to the scope stack rather than having a separate local scope.

I'm also curious about the alternative of not removing the current local scope if no scope is bound to the type.  What breaks if we do this instead?

**Comment by herbelin** [comment]
> Looking at the commits to HoTT: oh this is interesting. I think the thing I actually want here is to be able to bind category_scope to object Cat (while leaving object_scope bound to object _), and have bound scopes added to the scope stack rather than having a separate local scope.

I'm sorry @JasonGross, but I don't know how to help here. I don't know HoTT well enough to understand the ins and outs of a choice or another. Maybe could you just tell me if you're ok with the patch or if you'd prefer another one.

> I'm also curious about the alternative of not removing the current local scope if no scope is bound to the type. What breaks if we do this instead?

You mean as a general rule (rather than for the particular case of HoTT)?

Probably nothing would break further since this is the strategy applied up to now. I don't have a strong opinion. It looks a bit ad hoc to me but maybe it would be convenient in practice. I don't know if e.g. @ggonthier, who proposed the change, has an experience to share

**Comment by JasonGross** [comment]
@herbelin I'm willing to accept the patch.  I'd still be interested in the feature I mentioned; shall I open a new issue to request the ability to do things like `Bind Scope type_scope with (id Type).`?

**Comment by herbelin** [comment]
>  I'd still be interested in the feature I mentioned; shall I open a new issue to request the ability to do things like Bind Scope type_scope with (id Type).?

Well, yes. (And this reminds me of similar requests for coercions.)

And, on my side, I wonder whether there would be some need or support for an explicit syntax for temporary scope or to-stack scope in situations where the other semantics is taken. The current syntax either adds the scope to the stack (`term%scope` in an expression), or makes it a temporary scope (as in `Arguments foo x%scope.`). Would it be useful to provide say `term%%scope` to tell temporary in expressions and `Arguments foo x%!scope` to tell to stack?

In particular, this is obviously related to #3990, though I feel that the examples in #3990 are asking for more typing invariant being propagated, something that can be achieved more easily with typeclass-based notations. Being short in available time, I'm stopping there. BTW, it would be useful to have a comparison of the advantages and drawbacks of scope-based vs typeclass-based notations (or maybe such comparison already exists?).

**Comment by maximedenes** [comment]
> Note 2 : This PR depends on #6133 and #982 shall depend on it for interpreting casts in patterns.

Is it still the case?

**Comment by herbelin** [comment]
> Is it still the case? 

#6133 is closed and now part of this PR. For #982, I'll have to check but I suspect that it should be autonomous.

> What is the current status of this PR?

There are still a few incompatibilities (in HoTT and formal-topology), so I wonder whether @JasonGross' proposition of a mixed strategy (i.e. propagating the surrounding temporary scope if the type is not bound to a scope) could be a good compromise. Somehow, more user feedback would be useful. Maybe should we explicitly notify some people, e.g. those who participated to #3990, to take part to the discussion (and at the same time I know that everyone is very busy...).

In any case, my current view is to take a couple of days for letting mature what could be the optimal policy.

**Comment by SkySkimmer** [comment]
>In any case, my current view is to take a couple of days for letting mature what could be the optimal policy.

It has been a bit more than a couple days, any conclusions?

**Comment by herbelin** [comment]
> It has been a bit more than a couple days, any conclusions?

Hum, in the absence of alternative comments, I believe I should try with the "not removing the current local scope if no scope is bound to the type" patch we mentioned above.

**Comment by ejgallego** [comment]
Hi @herbelin , how can we make progress here? Maybe we could try merging some of the parts first?

**Comment by ejgallego** [comment]
Cleaning up my "assigned" list, please ping when activity resumes.

**Comment by ppedrot** [comment]
This PR has been lingering for a long time, with a "needs: progress" tag for almost two years. Judging from the context, it is unlikely to change in a near future, so I am willing to close it. @herbelin please reopen it when you're going to work on it.

**Comment by herbelin** [comment]
I believe it is worth being investigated further, but not a priority at the current time.

**Comment by proux01** [comment]
@herbelin the PR https://github.com/coq/coq/pull/18031 may help for the mathcomp failure

**Comment by herbelin** [comment]
@proux01 : If mathcomp is ok, yes, please proceed.

For the other failures, the PR could first be split into two:
- a first step to interpret only Prop, SProp, Set and Type (suspectingly fully compatible)
- a second step for using the scope binding (for which I already have overlays)

Incidentally, are the `: Z` casts important in the output number syntax test files?

**Comment by proux01** [comment]
> @proux01 : If mathcomp is ok, yes, please proceed.

@herbelin if you merge https://github.com/coq/coq/pull/18031 I can rebase this
 
> For the other failures, the PR could first be split into two:
> 
>     * a first step to interpret only Prop, SProp, Set and Type (suspectingly fully compatible)
> 
>     * a second step for using the scope binding (for which I already have overlays)

Well, if you already have the overlays, maybe it's easier to not split the PR?

> Incidentally, are the `: Z` casts important in the output number syntax test files?

No (I guess they were there to ensure the type of the parsed value but this is already ensured by the output), and considering they now break the test, I think they should just be removed.

**Comment by proux01** [comment]
@herbelin CI green here, I'd like to complete the doc, add a changelog entry and merge if you agree?

**Comment by herbelin** [comment]
> @herbelin CI green here, I'd like to complete the doc, add a changelog entry and merge if you agree?

OK on my side.

**Comment by proux01** [comment]
I'll merge by the end of next week if there are no more comments.

File: doc/changelog/03-notations/6134-master+type-scope-cast.rst [hunk 1]
```diff
@@ -0,0 +1,10 @@
+- **Changed:**
+  In casts like :g:`term : t` where :g:`t` is bound to some
+  scope :g:`t_scope`, via :cmd:`Bind Scope`, the :g:`term` is
```
Comment (jfehrle) [inline doc/changelog/03-notations/6134-master+type-scope-cast.rst:3]:
```suggestion
  scope :g:`t_scope`, via :cmd:`Bind Scope`, the :g:`term` is now
```

A little clearer about what has been changed.

File: doc/sphinx/user-extensions/syntax-extensions.rst [hunk 2]
```diff
@@ -1590,6 +1590,9 @@ Binding types or coercion classes to notation scopes
    then :g:`a` of type :g:`t` in :g:`f t a` is not recognized as an argument to
    be interpreted in scope ``scope``.
 
+   In explicit casts :g:`term : coercion_class`, the :g:`term`
+   will also be interpreted in :token:`scope_name`.
```
Comment (jfehrle) [inline doc/sphinx/user-extensions/syntax-extensions.rst:1594]:
```suggestion
   In explicit casts :n:`@term : @coercion_class`, the :n:`@term`
   is interpreted in the :token:`scope_name` associated with :n:`@coercion_class`.
```

:n: shows syntax with hyperlinks and in green
Seemed like "also" didn't add anything.

File: doc/sphinx/user-extensions/syntax-extensions.rst [hunk 3]
```diff
@@ -1590,6 +1590,9 @@ Binding types or coercion classes to notation scopes
...
+
```
Comment (jfehrle) [inline doc/sphinx/user-extensions/syntax-extensions.rst:1595]: If this applies to all casts (not just coercion_classes), then this should also be mentioned where `:>` is defined, i.e. https://coq.inria.fr/refman/language/core/definitions.html#type-cast

**Review by jfehrle — COMMENTED** [review]
Wording suggestions

File: doc/changelog/03-notations/6134-master+type-scope-cast.rst (hunk 1 above)
Comment (proux01) [reply doc/changelog/03-notations/6134-master+type-scope-cast.rst]: Thanks

File: doc/sphinx/user-extensions/syntax-extensions.rst (hunk 2 above)
Comment (proux01) [reply doc/sphinx/user-extensions/syntax-extensions.rst]: Thanks, I'm apparently not editing the doc often enough and forgot those things every time.

> Seemed like "also" didn't add anything.

Indeed, it's probably more confusing than useful.

File: doc/sphinx/user-extensions/syntax-extensions.rst (hunk 3 above)
Comment (proux01) [reply doc/sphinx/user-extensions/syntax-extensions.rst:1595]: Good point, done.

**Review by proux01 — APPROVED** [review]

**Comment by proux01** [comment]
Let's finally merge this almost six years old PR.

@coqbot merge now

**Pull request merged**